src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs

```diff
@@ -34,207 +44,169 @@ internal sealed class SlidingTextWindow : IDisposable
         /// </summary>
         public const char InvalidCharacter = char.MaxValue;
 
-        private const int DefaultWindowLength = 2048;
+        /// <summary>
+        /// This number was picked based on the roslyn codebase.  If we look at parsing all files,
+        /// this window ensures that 97.4% of all characters grabbed are within the window.  Note
+        /// that this is lower than expected, but heavily influenced by how heavily roslyn uses
+        /// large strings in its codebase to represent test content.  With all these files, lexemes
+        /// average around 60 characters.  If tests are excluded, then average lexeme length drops
+        /// to 25 characters, and the hit rate goes up to 98.4%.  Because this hitrate is so high,
+        /// we don't bother to dynamically resize the window.  Instead, when we hit the end, we just
+        /// grab in the next chunk of characters from the source text, as only one in every 65 lexemes
+        /// will need to go back to the source-text to read the full lexeme.
+        /// </summary>
+        public const int DefaultWindowLength = 4096;
+
+        private static readonly ObjectPool<char[]> s_windowPool = new ObjectPool<char[]>(() => new char[DefaultWindowLength]);
 
-        private readonly SourceText _text;                 // Source of text to parse.
-        private int _basis;                                // Offset of the window relative to the SourceText start.
-        private int _offset;                               // Offset from the start of the window.
-        private readonly int _textEnd;                     // Absolute end position
-        private char[] _characterWindow;                   // Moveable window of chars from source text
-        private int _characterWindowCount;                 // # of valid characters in chars buffer
+        /// <summary>
+        /// Underlying source text we are lexing.  This is the final truth of all characters.  Chunks of
+        /// this text will be read into a character window as needed to allow fast contiguous access to
+        /// a portion of the file at a time (as many <see cref="SourceText"/> implementations have logarithmic
+        /// access time to characters).
+        /// </summary>
+        public SourceText Text { get; }
 
-        private int _lexemeStart;                          // Start of current lexeme relative to the window start.
+        /// <summary>
+        /// Absolute end position of <see cref="Text"/>.  Attempts to read at or past this position will 
+        /// produce <see cref="InvalidCharacter"/>.
+        /// </summary>
+        private readonly int _textEnd;
```

Comment:
not sure how necessary this is. but given every read of a character compares current position to this, it seems reasonable to not go through a virt-call to _text.Length.